File: php/src/StatementPrinter.php
<?php

declare(strict_types=1);

namespace Theatrical;

use NumberFormatter;
use Theatrical\Pricing\PricingConfig;

class StatementPrinter
{
    private PricingConfig $pricingConfig;

    public function __construct(?PricingConfig $pricingConfig = null)
    {
        $this->pricingConfig = $pricingConfig ?? PricingConfig::fromFile(self::defaultConfigPath());
    }

    /**
     * @param array<string, Play> $plays
     */
    public function print(Invoice $invoice, array $plays): string
    {
        $totalAmount = Money::fromCents(0);
        $volumeCredits = 0;
        $format = new NumberFormatter($this->pricingConfig->currencyLocale(), NumberFormatter::CURRENCY);
        $currencyCode = $this->pricingConfig->currencyCode();

        $result = "Statement for {$invoice->customer}\n";

        foreach ($invoice->performances as $performance) {
            $play = $plays[$performance->playId];
            $calculator = PerformanceCalculator::create($performance, $play, $this->pricingConfig);
            $amount = $calculator->amount();

            $result .= "  {$play->name}: {$amount->format($format, $currencyCode)} ({$performance->audience} seats)\n";

            $totalAmount = $totalAmount->add($amount);
            $volumeCredits += $calculator->volumeCredits();
        }

        $result .= "Amount owed is {$totalAmount->format($format, $currencyCode)}\n";
        $result .= "You earned {$volumeCredits} credits";

        return $result;
    }

    private static function defaultConfigPath(): string
    {
        return __DIR__ . '/../config/pricing.json';
    }
}

File: java/src/main/java/theatricalplays/StatementPrinter.java
package theatricalplays;

import java.text.NumberFormat;
import java.util.Locale;
import java.util.Map;

public class StatementPrinter {

    private static final int CENTS_PER_DOLLAR = 100;

    public String print(Invoice invoice, Map<String, Play> plays) {
        var totalAmount = 0;
        var volumeCredits = 0;
        var result = new StringBuilder(String.format("Statement for %s%n", invoice.getCustomer()));

        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);

        for (var perf : invoice.getPerformances()) {
            var play = plays.get(perf.getPlayID());
            var calculator = PerformanceCalculator.create(perf, play);
            var thisAmount = calculator.amount();

            volumeCredits += calculator.volumeCredits();

            result.append(String.format("  %s: %s (%s seats)%n",
                    play.getName(), frmt.format(thisAmount / CENTS_PER_DOLLAR), perf.getAudience()));
            totalAmount += thisAmount;
        }
        result.append(String.format("Amount owed is %s%n", frmt.format(totalAmount / CENTS_PER_DOLLAR)));
        result.append(String.format("You earned %s credits%n", volumeCredits));
        return result.toString();
    }
}
